cuml/kmeans_demo_mnmg.ipynb

```diff
@@ -0,0 +1,236 @@
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# K-Means Multi-Node Mutli-GPU (MNMG) Demo\n",
+    "\n",
+    "K-Means multi-Node multi-GPU implementation leverages Dask to spread data and computations across multiple workers. cuML uses One Process Per GPU (OPG) layout, which maps a single Dask worker to each GPU.\n",
+    "\n",
+    "The main difference between cuML's MNMG implementation of k-means and the single-GPU is that the fit can be performed in parallel for each iteration, sharing only the centroids between iterations. The MNMG version also provides the same scalable k-means++ initialization algorithm as the single-GPU version.\n",
+    "\n",
+    "Unlike the single-GPU implementation, The MNMG k-means API requires a Dask cuDF Dataframe as input. `predict()` and `transform()` also return a Dask cuDF Dataframe. The Dask cuDF Dataframe API is very similar to the Dask DataFrame API, but underlying Dataframes are cuDF, rather than Pandas.\n",
+    "\n",
+    "For information on converting your dataset to Dask cuDF format: https://rapidsai.github.io/projects/cudf/en/latest/dask-cudf.html\n",
+    "\n",
+    "For additional information on cuML's MNMG k-means implementation: https://rapidsai.github.io/projects/cuml/en/latest/api.html#k-means-clustering"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import numpy as np\n",
+    "\n",
+    "import pandas as pd\n",
+    "import cudf as gd\n",
+    "\n",
+    "from cuml.test.dask.utils import dask_make_blobs\n",
+    "\n",
+    "from sklearn.metrics import adjusted_rand_score\n",
+    "\n",
+    "%matplotlib inline\n",
+    "import matplotlib.pyplot as plt\n",
+    "\n",
+    "from dask.distributed import Client\n",
+    "from dask_cuda import LocalCUDACluster\n",
+    "\n",
+    "from dask_ml.cluster import KMeans as skKMeans\n",
+    "from cuml.dask.cluster import KMeans as cumlKMeans"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Start Dask Cluster"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "cluster = LocalCUDACluster(threads_per_worker=1)\n",
+    "client = Client(cluster)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Define Parameters"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "n_samples = 1000000\n",
+    "n_features = 2\n",
+    "\n",
+    "n_total_partitions = len(list(client.has_what().keys()))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Generate Data\n",
+    "\n",
+    "### Device"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "X_df, X_cudf = dask_make_blobs(n_samples, \n",
```

Comment:
Would be nice to add a short text explanation here since I think persist is the one non-obvious bit to new users:

 - Generate random data on the workers and use `persist` to pin that data to each worker's active memory. (See https://distributed.dask.org/en/latest/memory.html for more information on `persist`)

Reply:
I've gotten in the habit of always calling persist when I create a dask dataframe. For one, it allows me to debug, by forcing the dataframe partitions to show up in calls to `client.has_what()` and `client.who_has()`. It's similar to Spark's `.cache()`. 

While it's not a bad idea to do this all the time, I suppose the example would be more clean and simple if we just removed them. What do you think?

Reply:
Also, we probably want to do the `client.sync` and explain why that's necessary for measuring timing of the algorithms.

Reply:
Further, rather than exposing the `extract_ddf_partitions` to the notebooks (which might be okay for 0.9), It might make more sense to add an argument `wait=True` to the dask_make_blobs function. I see this a lot in the Dask APIs.

Reply:
I don't see why we need extract_ddf_partitions here?  Could we do client.wait(X_cudf, X_df) instead?

Reply:
Unfortunately client does not have a `wait` function, but it does look like `dask.distributed.wait` works. Not sure why I didn't think that worked on delayed objects. I fixed this in our shared branch. Once the tests finish, I'll push it.